subworkflows/local/singular_demultiplex/main.nf

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env nextflow
+
+//
+// Demultiplex Singular Genomics bases data using sgdemux
+//
+
+include { SGDEMUX } from "../../../modules/nf-core/sgdemux/main"
+
+workflow SINGULAR_DEMULTIPLEX {
+    take:
+        // Input fastq's must be bgzipped for compatibility with sgdemux
+        // samplesheet.csv must be a two column csv = Sample_Barcode,Sample_ID)
+        ch_flowcell     // [[id:"", lane:""],samplesheet.csv, path/to/bases/files]
+
+    main:
+
+        // MODULE: sgdemux
+        // Demultiplex the bases files
+        SGDEMUX( ch_flowcell )
+
+        // Generate meta for each fastq
+        ch_fastq_with_meta = generate_fastq_meta(SGDEMUX.out.sample_fastq)
+
+    emit:
+        fastq                   = ch_fastq_with_meta
+        //TODO: @samfulcrum veriify outputs are correct compared to what is in bases2fastq
+        generated_run_manifest  = SGDEMUX.out.per_project_metrics 
+        metrics                 = SGDEMUX.out.metrics 
+        unassigned              = SGDEMUX.out.most_frequent_unmatched 
+        versions                = SGDEMUX.out.versions
+}
+
+/*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    FUNCTIONS
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+*/
+
+// Add meta values to fastq channel
+def generate_fastq_meta(ch_reads) {
+    // Create a tuple with the meta.id and the fastq
+    ch_reads.transpose().map{
+        fc_meta, fastq ->
+        def meta = [
+            "id": fastq.getSimpleName().toString() - ~/_R[0-9].*$/,
+            "samplename": fastq.getSimpleName().toString() - ~/_R[0-9].*$/,
+            "readgroup": [:],
+            "fcid": fc_meta.id,
+            "lane": fc_meta.lane
+        ]
+        meta.readgroup = readgroup_from_fastq(fastq)
+        meta.readgroup.SM = meta.samplename
+
+        return [ meta , fastq ]
+    }
+    // Group by meta.id for PE samples
+    .groupTuple(by: [0])
+    // Add meta.single_end
+    .map {
+        meta, fastq ->
+        if (fastq.size() == 1){
+            meta.single_end = true
+        } else {
+            meta.single_end = false
+        }
```

Comment:
I'd be curious what the demultiplex pipeline's intention is here.  We could have up to four FASTQs (I1, I2, R1, R2) and single end could have there (I1, I2, R1) and so not sure if this is the best heuristic.  Thoughts?

Reply:
That's just a common nf-core pattern. So you're right, probably a little naive here, so any ideas on how else to do it?

Reply:
I am not sure why it should be inferred at all and matter for demultiplexing.  At least for this demux tool, it can accept one or more FASTQs, even for a given read type (e.g. index reads or first end of a pair) so that you can demux all the lanes at a single time.  Perhaps other tools (e.g. `fastp`) needed it?  Even inferring it from the # of output files produced for a single sample is fraught, since the `sgdemux` tool can output separate FASTQs for index reads (and I think molecular barcodes if you don't want them in the read name/header).  You'd have to search for the number of output FASTQs per sample that have either `R1` or `R2`.

For `sgdemux`, we use a [read structure](https://github.com/fulcrumgenomics/fgbio/wiki/Read-Structures) to assign reads to their logical components.  The read structure may not be known ahead of time, since it could be inferred from the FASTQ files present on disk.

Perhaps rather than inferring, it should be a required option to the pipeline.

Reply:
@nils sounds like a plan, thank you for the detail.